### PR TITLE
Rework mod object to be the intended way to interact with systems

### PR DESCRIPTION
**Pull request by Enduriel**
SystemModAddons are added to a mod when the mod registers using `mod.register(_system)`, these are then using to interact with systems using `mod.SystemName.function()`.

A `EmptyModAddon` is added to be placeholder for `SystemModAddons` before they are initialized to throw a more useful error when a user tries to use a function from a mod system without registering.

### CODE REVIEW

File: msu/systems/debug/debug_mod_addon.nut [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	{
+		::MSU.System.Debug.setFlag(this.getID(), _flagID, _flagBool);
+	}
+
+	function isEnabled( _flagID = "default" )
+	{
+		::MSU.System.Debug.isEnabledForMod(this.getID(), _flagID);
+	}
+
+	function print( _text, _logType, _flagID = "default" )
```
Comment (TaroEld) [inline msu/systems/debug/debug_mod_addon.nut:18]: This seems to lack the printWarning etc helper functions

File: msu/systems/keybinds/key_static.nut [hunk 2]
```diff
@@ -30,7 +30,7 @@ this.MSU.Key <- {
 		if (keyArray.len() > 1)
 		{
 			keyArray.sort();
-			key = keyArray.reduce(@(a, b) a + "+" + b) + "+";
+			key = keyArray.reduce(@(_a, _b) _a + "+" + _b) + "+";
```
Comment (TaroEld) [inline msu/systems/keybinds/key_static.nut:33]: Lambda function parameters traditionally don't get a _ prefix, I'm afraid you'll have to update your syntax highlighter

File: msu/systems/keybinds/keybinds_mod_addon.nut [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	function addSQKeybind( _id, _keyCombinations, _state, _function, _name = null, _keyState = null, _description = "" )
+	{
+		local keybind = ::MSU.Class.KeybindSQ(this.getID(), _id, _keyCombinations, _state, _function, _name, _keyState);
+		keybind.setDescription(_description);
+		::MSU.System.Keybinds.add(keybind);
+	}
+
+	function addJSKeybind( _id, _keyCombinations, _name = null, _keyState = null, _description = "" )
+	{
+		local keybind = ::MSU.Class.KeybindJS(this.getID(), _id, _keyCombinations, _name, _keyState);
```
Comment (TaroEld) [inline msu/systems/keybinds/keybinds_mod_addon.nut:22]: Needs to be updated to use the new JS method as that doesn't take keystate

File: msu/systems/mod.nut [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+	constructor( _id, _version, _name = null )
+	{
+		if (_name == null) _name = _id;
+		this.MSU.requireString(_id, _version, _name);
+
+		this.ID = _id;
+		this.Name = _name;
+
+		local table = ::MSU.System.Registry.getVersionTable(_version);
```
Comment (TaroEld) [inline msu/systems/mod.nut:24]: This should probably handled via the Serialization modAddon

File: msu/systems/mod.nut [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		this.Debug.print(_text, ::MSU.System.Debug.LogType.Error, _flagID);
+	}
+
+	// No isDebugEnabled cuz imo Debug.isEnabled is fine
+
+	function getSetting( _id )
+	{
+		return this.ModSettings.getSetting(_id);
+	}
+}
```
Comment (TaroEld) [inline msu/systems/mod.nut:100]: Add a _tostring method to print out all the relevant factors of the mod?

File: msu/systems/mod_settings/load.nut [hunk 6]
```diff
@@ -122,4 +123,4 @@ logPage.add(verboseModeToggle);
 
 
 // Testing Code
-testSettingsSystem();
```
Comment (TaroEld) [inline msu/systems/mod_settings/load.nut:125]: This should be guarded by MSU.Mod.isEnabled("modsettings") instead

File: msu/systems/serialization/load.nut [hunk 7]
```diff
@@ -3,5 +3,7 @@ this.MSU.System.Serialization <- this.MSU.Class.SerializationSystem();
 
 ::isSavedVersionAtLeast <- function( _modID, _version )
```
[the remark this answers is not in the document]
Comment (TaroEld) [reply msu/systems/serialization/load.nut:4]: I would say so

File: msu/systems/keybinds/key_static.nut (hunk 2 above)
Comment (Enduriel) [reply msu/systems/keybinds/key_static.nut:33]: _No_

File: msu/systems/keybinds/keybinds_mod_addon.nut (hunk 3 above)
Comment (Enduriel) [reply msu/systems/keybinds/keybinds_mod_addon.nut:22]: Neither of them is merged in yet, I've mentioned this on the other PR as I was hoping to merge this one in first.

File: msu/systems/mod_settings/load.nut (hunk 6 above)
Comment (Enduriel) [reply msu/systems/mod_settings/load.nut:125]: Unsure what you mean here

File: msu/systems/mod.nut (hunk 5 above)
Comment (Enduriel) [reply msu/systems/mod.nut:100]: Good call

**Pull request merged**